nx/test/nx/defn/grad_test.exs

```diff
@@ -1646,6 +1646,20 @@ defmodule Nx.Defn.GradTest do
     end
   end
 
+  describe "qr" do
+    defn qr_grad(t) do
+      grad(t, fn tensor ->
+        {q, r} = Nx.LinAlg.qr(tensor)
+
+        Nx.sum(Nx.concatenate([q, r]))
+      end)
+    end
+
+    test "computes grad for tensor" do
+      assert qr_grad(Nx.tensor([[1.0, 2.0], [1.0, -1.0]])) == Nx.tensor([[1, 1], [-1, 1]])
+    end
```

Comment:
This test is giving me a weird exception which I suspect has to do with `{q, r} = ...`

```elixir
  1) test qr computes grad for tensor (Nx.Defn.GradTest)
     test/nx/defn/grad_test.exs:1658
     ** (ArgumentError) argument error
     code: assert qr_grad(Nx.tensor([[1.0, 2.0], [1.0, -1.0]])) == Nx.tensor([[1, 1], [-1, 1]])
     stacktrace:
       :erlang.element(1, #Nx.Tensor<
  f32[2][2]
  
  Nx.Defn.Expr
  parameter a                                f32[2][2]
  tensor g                                   u8[2][2]
  tensor o                                   s64[2][2]
  tensor q                                   s64[2][2]
  tensor r                                   s64[2][2]
  tensor bd                                  s64[2][2]
  b = metadata [ a, [Nx.Defn.Grad] ]         f32[2][2]
  c = slice [ 1.0, [0, 0], [2, 2], [1, 1] ]  f32[2][2]
  d = multiply [ b, c ]                      f32[2][2]
  e = metadata [ d, [Nx.Defn.Grad] ]         f32[2][2]
  f = add [ e, e ]                           f32[2][2]
  h = multiply [ b, c ]                      f32[2][2]
  i = metadata [ h, [Nx.Defn.Grad] ]         f32[2][2]
  j = transpose [ i, [1, 0] ]                f32[2][2]
  k = dot [ b, [1], j, [0] ]                 f32[2][2]
  l = transpose [ e, [1, 0] ]                f32[2][2]
  m = dot [ l, [1], b, [0] ]                 f32[2][2]
  n = subtract [ k, m ]                      f32[2][2]
  p = select [ g, n, o ]                     f32[2][2]
  s = select [ q, n, r ]                     f32[2][2]
  t = add [ p, s ]                           f32[2][2]
  u = transpose [ t, [1, 0] ]                f32[2][2]
  v = add [ u, p ]                           f32[2][2]
  w = dot [ f, [1], v, [0] ]                 f32[2][2]
  x = transpose [ b, [1, 0] ]                f32[2][2]
  y = transpose [ x, [1, 0] ]                f32[2][2]
  z = qr [ {y, y}, y, mode: :reduced ]       tuple2
  ba = elem [ z, 0, 2 ]                      f32[2][2]
  bb = elem [ z, 1, 2 ]                      f32[2][2]
  bc = transpose [ bb, [1, 0] ]              f32[2][2]
  be = triangular_solve [ bc, bd, [] ]       f32[2][2]
  bf = dot [ ba, [1], be, [0] ]              f32[2][2]
  bg = transpose [ bf, [1, 0] ]              f32[2][2]
  bh = dot [ w, [1], bg, [0] ]               f32[2][2]
  bi = metadata [ bh, [Nx.Defn.Grad] ]       f32[2][2]
>)
       (nx 0.1.0-dev) lib/nx/defn/grad.ex:234: Nx.Defn.Grad.grad/5
       (nx 0.1.0-dev) lib/nx/defn/grad.ex:161: anonymous fn/3 in Nx.Defn.Grad.to_grad/3
       (elixir 1.11.2) lib/enum.ex:1521: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (nx 0.1.0-dev) lib/nx/defn/grad.ex:195: Nx.Defn.Grad.grad_pairs/3
       (nx 0.1.0-dev) lib/nx/defn/grad.ex:161: anonymous fn/3 in Nx.Defn.Grad.to_grad/3
       (nx 0.1.0-dev) lib/nx/defn/grad.ex:24: Nx.Defn.Grad.transform/2
       (nx 0.1.0-dev) lib/nx/defn/kernel.ex:232: Nx.Defn.Kernel.grad/2
       test/nx/defn/grad_test.exs:1650: anonymous fn/3 in Nx.Defn.GradTest.qr_grad/1
       (nx 0.1.0-dev) lib/nx/defn/evaluator.ex:21: Nx.Defn.Evaluator.__jit__/4
       test/nx/defn/grad_test.exs:1659: (test)
```

Keep in mind that since `Nx.LinAlg.invert` is Nx-native, its underlying expressions are all present in the expr's introspection

Reply:
@josevalim do you have any idea on what might be going on here? This is the same error I mentioned the other day

Reply:
I believe this is because `qr` is meant to return a tuple. So the grad of `qr` has to return a tuple but it is currently returning a single element. For example, you could make it so it returns this:

```
{res, cache} = to_grad(input, da, cache)
{{res, res}, cache}
```

but i don't know if this is the correct result. :)

Reply:
Thanks! I'll look into this and see what it yields

Reply:
This removed the exception, although the result isn't correct, but this might be a problem in the calculation itself

Reply:
For reference, this is Jax's implementation:
```python
def qr_jvp_rule(primals, tangents, full_matrices):
  # See j-towns.github.io/papers/qr-derivative.pdf for a terse derivation.
  x, = primals
  dx, = tangents
  q, r = qr_p.bind(x, full_matrices=False)
  *_, m, n = x.shape
  if full_matrices or m < n:
    raise NotImplementedError(
      "Unimplemented case of QR decomposition derivative")
  dx_rinv = triangular_solve(r, dx)  # Right side solve by default
  qt_dx_rinv = jnp.matmul(_H(q), dx_rinv)
  qt_dx_rinv_lower = jnp.tril(qt_dx_rinv, -1)
  do = qt_dx_rinv_lower - _H(qt_dx_rinv_lower)  # This is skew-symmetric
  # The following correction is necessary for complex inputs
  do = do + jnp.eye(n, dtype=do.dtype) * (qt_dx_rinv - jnp.real(qt_dx_rinv))
  dq = jnp.matmul(q, do - qt_dx_rinv) + dx_rinv
  dr = jnp.matmul(qt_dx_rinv - do, r)
  return (q, r), (dq, dr)
```

I'm not sure if this is mathematically equivalent to the definition I was using, but I'll aim for this implementation instead

Reply:
@josevalim I've basically ported line by line this definition, with the difference that because Jax jvps receive the grad, I had to use `grad/5` instead. However, the result isn't what is expected.

Any ideas on what where the issue could be?